[Reservation, Cart] 예약 및 장바구니 기능 구현

J-Cheol opened the pull request:
- close #48 

### 작업목록
- [x] 예약 정보 저장 필드(name, phone) 삭제
- [x] 예약 정보 저장 필드(member_id) 생성 -> 연관관계 설정 -> 단방향(추후 필요할때 양방향 설정)
- [x] 필드 변경에 따른 로직 변경
- [x] Swagger 문서 변경
- [x] 장바구니 연관관계 설정
- [x] 장바구니 저장 기능 구현
- [x] 장바구니 조회 기능 구현
- [x] 장바구니 삭제 기능 구현
- [x] 장바구니 옵션 수정 기능 구현
- [x] principal 적용
- [x] 결제정보를 위한 회원 정보 기능 구현
- [x] 장바구니 -> 결제 -> 예약테이블에 저장하는 기능 구현
- [x] 예약접수시 solapi 연동(비동기처리)
- [x] Cart 리팩토링

### 참고사항
ReservationService의 경우 현재 리팩토링 진행을 하지 않았습니다.
다음 스프린트 기간에 맞춰 리팩토링 진행예정인점 참고해주세요.

스프린트 기간중 건드린 부분이 많습니다.
장바구니 관련된 로직 리팩토링이 완료되었으나 다른 부분이 미흡할 수 있습니다.
확인 후 리뷰 부탁드려요! 감사합니다.

2024 / 12 / 15 : reservation <-> cart 도메인 분리

ppusda committed:
build: Spring Security 추가 및 빌드 파일 정리

ppusda committed:
chore: 설정 파일 업데이트

ppusda committed:
feat: Spring Security 설정 및 JWT Filter, Provider 정의

ppusda committed:
feat: 회원 Entity 추가

ppusda committed:
feat: 회원 로그인 기능 구현

ppusda committed:
feat: 토큰 Entity 추가

ppusda committed:
feat: 토큰 관리 기능 구현

ppusda committed:
feat: ToucheeseTokenInvalidException, ToucheeseUnAuthorizedException 추가

ppusda committed:
feat: 요청 API 허용 설정 추가

ppusda committed:
comment: 메서드 주석 추가

ppusda committed:
refactor: token expiration 설정 파일로 분리

ppusda committed:
refactor: accessToken 갱신 시 응답 헤더에 포함

ppusda committed:
refactor: 응답 형식 통합 및 정형화 (ApiResponse 추가)

ppusda committed:
Merge branch 'dev' of https://github.com/Mac-Win/Toucheese-BackEnd into 41-로그인-기능-구현

ppusda committed:
refactor: username -> email로 변경

ppusda committed:
refactor: 토큰 검증 로직 변경
 - ACCEPTED, EXPIRED, DENIED로 분류
 - 만료 시 헤더로 재갱신 된 토큰 전달
 - 검증과 파싱 분리

ppusda committed:
merge: 추가 수정사항 반영 및 충돌 해결

ppusda committed:
refactor: 로그인 시 응답 형식 변경

ppusda committed:
Merge branch 'dev' of https://github.com/Mac-Win/Toucheese-BackEnd into 41-로그인-기능-구현

ppusda committed:
refactor: addAllowedOrigin에 개발환경 추가

ppusda committed:
refactor: CORS 설정 추가 / Authorization 헤더 접근 허용

J-Cheol committed:
feat : 이름, 전화번호 필드 -> MemberId로 변경

J-Cheol committed:
feat : 장바구니 기능 구현 - Create

장바구니 Create 기능 구현
Swagger 설명 추가

J-Cheol committed:
feat : 장바구니 목록 조회 기능 구현

J-Cheol committed:
feat : 장바구니 삭제 기능 구현, 엔드포인트 변경

J-Cheol committed:
feat : 장바구니 옵션, 인원 업데이트 기능 구현

J-Cheol committed:
refactor : PatchMapping -> PutMapping으로 변경 장바구니

J-Cheol committed:
feat : principal 기능 추가

J-Cheol committed:
Merge branch 'dev' of https://github.com/Mac-Win/Toucheese-BackEnd into 48-예약-및-장바구니-기능-구현-회원적용

J-Cheol committed:
feat : SwaggerConfig 추가, yml 분리

J-Cheol committed:
chore :  Change active Spring profile from dev to prod

J-Cheol committed:
chore: Add application-secret.yml to .gitignore

J-Cheol committed:
feat : 장바구니 업데이트 기능 principal 적용

해당 회원만 장바구니 업데이트 가능

J-Cheol committed:
feat : CartResponse cartId 추가

J-Cheol committed:
docs : 필드추가에 따른 swagger 수정 getCartList

J-Cheol committed:
feat : 회원 정보 기능 구현

J-Cheol committed:
feat : 장바구니 리스트 조회 시 상품 이미지 필드 추가

J-Cheol committed:
feat : 장바구니 리스트 studioImage 필드 추가

J-Cheol committed:
feat : 장바구니 저장 기능 memberId를 principal로 구현

J-Cheol committed:
feat : 장바구니 조회 기능중 옵션 부분 추가 구현

장바구니 옵션 변경할때 기준이 되는 상품 정보가 없기에 추가로 넣었습니다.

J-Cheol committed:
feat : 상품 가격 필드 추가

J-Cheol committed:
feat : toCsv null or 비었을 경우 빈 값 출력

J-Cheol committed:
feat : 상품 결제 조회화면 및 장바구니 데이터를 예약 테이블로 복사 후 삭제 기능 구현

장바구니 데이터를 예약테이블로 복사 후 삭제하는 기능 구현

J-Cheol committed:
Merge branch 'dev' of https://github.com/Mac-Win/Toucheese-BackEnd into 48-예약-및-장바구니-기능-구현-회원적용

J-Cheol committed:
feat : 결제 완료시 문자 + 이메일 기능 연동 구현

J-Cheol committed:
Merge branch 'dev' of https://github.com/Mac-Win/Toucheese-BackEnd into 48-예약-및-장바구니-기능-구현-회원적용

J-Cheol committed:
feat : reservationCompletedAt 타임존 서울 설정

J-Cheol committed:
build : mail 보내기 관련 의존성 추가

J-Cheol committed:
feat : LocalTime 초단위 삭제

J-Cheol committed:
feat : 예약 관련 회원 인증 기능 추가

isAuthenticated() 추가

J-Cheol committed:
feat : 결제 완료시 문자, 이메일 전송 비동기처리

J-Cheol committed:
build : tx 의존성 추가 / refactor : 현재 사용하지 않는 기능 삭제 후 최신화

J-Cheol committed:
refactor : MemberService MemberId 가져오는 메서드 생성

isAuthenticated()가 적용되어있는 인증된 회원 id를 가져옵니다.

J-Cheol committed:
chore : MySQL 버전 명시 제거

J-Cheol committed:
refactor : CsvUtils 유틸 인스턴스화 방지 및 from 공백문자 제거 기능 추가

J-Cheol committed:
refactor : 장바구니 결제 완료시 예약 접수 메서드 리팩토링

1. Controller에 맞는 역할로 refactoring 진행
2. Service 로직 간소화
3. solapi와 직접적인 연관을 끊고 이벤트리스너를 통해 접근

J-Cheol committed:
refactor : 장바구니 저장 - 역할에 따라 계층별로 로직을 분리

J-Cheol committed:
refactor : hibernate.dialect 설정 제거

J-Cheol committed:
refactor : 장바구니 조회 - 서비스 로직 간소화, 재사용성 및 가독성을 위해 Helper메서드로 분리

J-Cheol committed:
refactor : 장바구니 삭제 - 중복 로직 삭제 및 검증 메서드명 변경

J-Cheol committed:
feat : 장바구니 삭제 기능 추가 - 다중선택 삭제 기능 구현

J-Cheol committed:
refactor : 장바구니 업데이트 로직 간소화, 중복 메서드 제거

J-Cheol committed:
refactor : 선택한 장바구니 상품 조회 (회원, 선택한상품들)

J-Cheol committed:
refactor : 결제완료 status를 예약접수로 변경

ppusda commented on src/main/java/com/toucheese/member/service/MemberService.java:
```diff
@@ -54,4 +58,21 @@ private void checkMemberPassword(Member member, String password) {
+     * principal MemberId 찾는 메서드
+     * PreAuthorize("isAuthenticated()")가 적용되어 있을 경우
+     */
+    @Transactional(readOnly = true)
```
이 부분에 `@Transactional`을 사용하신 이유가 있으신가요?

J-Cheol replied:
여기에 필요 없는 부분인데 리팩토링 과정에서 잘못생각하고 넣었습니다. 머슥

J-Cheol committed:
Merge branch 'dev' of https://github.com/Mac-Win/Toucheese-BackEnd into 48-예약-및-장바구니-기능-구현-회원적용

J-Cheol committed:
refactor : Valid 어노테이션 사용하였으나 적용 안시킨점 수정

MemberId같은 경우 현재 Principal 사용으로 인해 불필요하지만 처음 작성한 부분으로 인해 안드로이드와 웹에서 수정해야하는 점 때문에 그대로 남겨두었습니다.

ppusda commented on src/main/java/com/toucheese/solapi/service/MessageService.java:
```diff
@@ -22,20 +24,32 @@ public class MessageService {
+        try {
+            sendSms(member.getPhone(), messageText);
+        } catch (Exception e) {
+            System.err.println("Failed to send SMS: " + e.getMessage());
```
로깅 하실 때는 `@Slf4j`를 사용하시면 될 것 같습니다!
로깅을 좀 더 편하고 안정적으로 할 수 있습니당.

```java
try {
    sendSms(member.getPhone(), messageText);
} catch (Exception e) {
     log.error("Failed to send SMS: {}", e.getMessage());
}
```

J-Cheol replied:
SLF4J 적용해보도록 하겠습니다

ppusda commented on src/main/java/com/toucheese/member/service/MemberService.java:
```diff
@@ -54,4 +58,21 @@ private void checkMemberPassword(Member member, String password) {
+     */
+    @Transactional(readOnly = true)
+    public Long getAuthenticatedMemberId(Principal principal) {
+        return Long.parseLong(principal.getName());
```
단순히 `Long.parseLong(principal.getName());`만 수행하는 거면 각 로직에서 수행할 필요없이 컨트롤러 단에서 던져주는게 더 좋아보입니다!

J-Cheol replied:
처음에는 컨트롤러 단에서 던져주었으나 컨트롤러는 받은 값 그대로 넘겨줘야한다고 생각하여 CartService에 넣었었습니다. 그리고 해당 부분이 Member와 관련된 부분이다보니 저렇게 생각이 흘러갔네요.
리뷰에서 말씀해주신것처럼 다시 생각해보니 간단한 작업이라 컨트롤러에서 작업하는게 좋을 것 같습니다.
서비스는 비즈니스 로직을 수행해야하는데 과했던 것 같습니다.

ppusda commented on src/main/java/com/toucheese/member/service/MemberService.java:
```diff
@@ -54,4 +58,21 @@ private void checkMemberPassword(Member member, String password) {
+    public MemberContactInfoResponse findMemberContactInfo(Principal principal) {
+        Long memberId = getAuthenticatedMemberId(principal);
+
+        Member member = memberRepository.findById(memberId)
```
위쪽에 생성해 둔 `findMemberById`가 있으니 그걸로 대체해도 될 것 같습니다!

J-Cheol replied:
감사합니다. 체크를 못했네요!

ppusda commented on src/main/java/com/toucheese/reservation/service/CartService.java:
```diff
@@ -0,0 +1,234 @@
+	private final CartRepository cartRepository;
+	private final StudioService studioService;
+	private final ProductService productService;
+	private final MemberService memberService;
```
아무래도 `getAuthenticatedMemberId`를 처리하기 위해서 목적이 더 강해져보이는데 코드를 살펴보다보니 memberService에서 해당 로직을 담당하는 게 조금 부자연스럽기도 하며 `MessageController`에도 동일한 로직이 있어 Util 클래스로 새로 생성하는게 좋을 것 같습니다.

```Java
public class PrincipalUtils {
    public static Long extractMemberId(Principal principal) {
        try {
            return Long.parseLong(principal.getName());
        } catch (NumberFormatException e) {
            throw new ToucheeseUnAuthorizedException("Principal name is not a valid memberId format: " + principal.getName());
        }
    }
}
```

J-Cheol replied:
아마 MessageController는 유진님께 물어보고 삭제할 것 같습니다.
getAuthenticatedMemberId가 너무 과하기도 하고 Service에서는 비즈니스 로직만을 수행해야하는데 불필요한 점 인지하였습니다.

junggg097 replied:
Principal 객체에서 memberId 를 추출하는 로직 중복 되어 있어서 , Util 클래스로 생성해서 공통으로 사용하는 쪽으로 변경하는 게 좋을 거 같네용

J-Cheol replied:
처음에는 유틸로 따로 추출할까도 생각했습니다.
하지만 다른 컨트롤러에서 사용하지 않고 ReservaionController에서만 진행되기때문에 빼지 않았습니다.
찾아보았을때 단순한 변환이거나 타 도메인의 Controller나 Service에서 사용된다면 유틸로 추출하는게 좋으나 아니라면 컨트롤러 내부에서 처리하는게 낫다는걸 보고 진행하였습니다.

*회원이 관련된 경우 재사용빈도가 높은것으로 판단되어 유틸로 빼겠습니다!*

J-Cheol committed:
refactor : getAuthenticatedMemberId제거, Controller에서 MemberId 타입 Long으로 변경, 중복 로직 간소화

J-Cheol committed:
refactor : Lombok Slf4J 사용

J-Cheol committed:
refactor : 중복 코드 유틸화

J-Cheol committed:
refactor : 장바구니, 예약 컨트롤러 분리

J-Cheol committed:
refactor : reservation <-> cart 도메인 분리

ppusda commented on src/main/java/com/toucheese/product/service/ProductService.java:
```diff
@@ -49,4 +52,9 @@ public Product findProductById(Long productId) {
         return productAddOptionRepository.findByProduct_IdAndAddOption_IdIn(productId, addOptionIds);
     }
+
+    public AddOption findAddOptionById(Long addOptionId) {
```
`@Transactional` 추가해주세욥!

J-Cheol replied:
아 여기에 추가하고 인지를 못했네요 감사합니다

ppusda commented on src/main/java/com/toucheese/cart/service/CartService.java:
```diff
@@ -0,0 +1,227 @@
+	 * 장바구니 소유권 검증 및 Cart 조회
+	 */
+	private Cart validateCartOwnership(Long cartId, Long memberId) {
+		Cart cart = cartRepository.findById(cartId)
```
이 부분도 따로 메서드로 분리하여 Transactional 붙여주시면 될 것 같습니다!

J-Cheol replied:
해당 부분은 self-invocation문제로 인하여 저렇게 사용했었습니다.
CartReadService를 만들어서 처리하도록 하겠습니다.

ppusda commented on src/main/java/com/toucheese/cart/repository/CartRepository.java:
```diff
@@ -0,0 +1,22 @@
+
+	List<Cart> findByMember(Member member);
+
+	@Query("SELECT c.member FROM Cart c WHERE c.id = :cartId")
```
아래 메서드들은 네이밍으로도 같은 동작을 유도할 수 있을 것 같은데, JPQL을 사용하신 이유가 있으신가요?

J-Cheol replied:
동국님 말씀이 맞습니다.
가독성을 위해서 사용하였었는데 JPQL을 사용하니 조금 가독성이 떨어지는것 같네요.
해당 부분은 현재 사용하고 있지 않아서 삭제하겠습니다.

J-Cheol committed:
refactor : CartReadService 생성 - self-invocation 해결, 메서드 분리

J-Cheol committed:
refactor : 사용하지 않는 코드 삭제

J-Cheol committed:
refactor : JPQL 삭제

ppusda reviewed (approved):
고생많으셨습니다!
